symbolic-unwind/src/evaluator/mod.rs

```diff
@@ -146,72 +168,141 @@ impl<'memory, A: RegisterValue, E: Endianness> Evaluator<'memory, A, E> {
                     .memory
                     .as_ref()
                     .ok_or(EvaluationError(EvaluationErrorInner::MemoryUnavailable))?;
-                memory.get(address, self.endian).ok_or_else(|| {
-                    EvaluationError(EvaluationErrorInner::IllegalMemoryAccess {
+                if let Some(val) = memory.get(address, self.endian) {
+                    val
+                } else {
+                    return Err(EvaluationError(EvaluationErrorInner::IllegalMemoryAccess {
                         address: address.try_into().ok(),
                         bytes: A::WIDTH,
                         address_range: memory.base_addr..memory.base_addr + memory.len() as u64,
-                    })
-                })
+                    }));
+                }
+            }
+        };
+        Ok(val)
+    }
+
+    /// Evaluates the given register's rule and returns the value.
+    ///
+    /// This fails if there is no rule for the register or it cannot be evaluated.
+    /// Results are cached.
+    pub fn evaluate_register(&mut self, register: &Register) -> Result<A, EvaluationError> {
+        if let Some(val) = self.register_cache.get(register) {
+            return Ok(*val);
+        }
+
+        if *register == Register::cfa() {
+            let cfa_rule = match self.cfa_rule.take() {
+                Some(e) => e,
+                None => {
+                    return Err(EvaluationError(EvaluationErrorInner::NoRuleForRegister(
+                        register.clone(),
+                    )))
+                }
+            };
+            let result = self.evaluate(&cfa_rule);
+            self.cfa_rule = Some(cfa_rule);
+            if let Ok(val) = result {
+                self.register_cache.insert(Register::cfa(), val);
             }
+            result
+        } else {
+            let e = match self.cfi_rules.remove(register) {
```

Comment:
You `remove`, `evaluate`, `insert`. Its not really obvious why that is needed. Would be nice to either document this or find a solution that doesn’t need that. Same for the `take`.

Reply:
It's a borrow checker issue. I need to take the rule I'm considering out of the evaluator because I want to call `evaluate` with it, which borrows mutably. So at that point I can't have a borrow of the evaluator.